### PR TITLE
fix variant initialization logic to allow proper handling of values="*"

### LINKED ISSUE #40405: Variant.__init__ logic error when values == '*'
### Steps to reproduce

When setting `variant(name, default, description, values="*")`, [`spack.variant.Variant` sets `self.single_value_validator = lambda x: True`](https://github.com/spack/spack/blob/6c7b2e1056bbd158544da9b6a79fd4099c18cd7a/lib/spack/spack/variant.py#L62).  But, logic block started on [line 77](https://github.com/spack/spack/blob/6c7b2e1056bbd158544da9b6a79fd4099c18cd7a/lib/spack/spack/variant.py#L77)  overwrites whatever `self.single_value_validator` was previously set to.  The solution appears to be replacing `if callable` with `elif callable`.

### Error message

No particular error message, but the `"*"` wildcard is not obeyed.

### Information on your system

System independent.

### General information

- [X] I have run `spack debug report` and reported the version of Spack/Python/Platform
- [X] I have searched the issues of this repo and believe this is not a duplicate
- [X] I have run the failing commands in debug mode and reported the output

### PR DESCRIPTION
**Pull request by tjfulle**
Fix for #40405.

Closes: #40405

### CODE REVIEW

**Review by alalazo — CHANGES_REQUESTED** [review]
We need a regression test for this fix. Also, we should double check possible edge cases and differences between `values=str` and `values=*` - in case they should behave the same it's maybe preferable to have the same validator for both.

**Comment by psakievich** [comment]
@alalazo I put together a quick test to compare the resulting differences between `str` and `"*"`.  The end result looks equivalent to me. Not sure if this is sufficient, or if there are additional corner cases you want to evaluate as part of this PR.

**Comment by psakievich** [comment]
@spackbot fix style

File: lib/spack/spack/test/variant.py [hunk 1]
```diff
@@ -734,3 +734,38 @@ def test_conditional_value_comparable_to_bool(other):
...
+    wild_var.validate_or_raise(wild_output)
+    # str case
+    str_output = str_var.make_variant(several_arbitrary_values)
+    str_var.validate_or_raise(str_output)
+    # swap and validate outputs
+    # must swap names to ensure the contents are the same
+    str_output.name = wild_var.name
+    wild_output.name = str_var.name
+    str_var.validate_or_raise(wild_output)
+    wild_var.validate_or_raise(str_output)
```
Comment (alalazo) [inline lib/spack/spack/test/variant.py]: Can't we use assertion here, instead of patching the objects?

File: lib/spack/spack/test/variant.py [hunk 2]
```diff
@@ -734,3 +734,38 @@ def test_conditional_value_comparable_to_bool(other):
     value = spack.variant.Value("98", when="@1.0")
     comparison = value == other
     assert comparison is False
+
+
+@pytest.mark.regression("40405")
+def test_wild_card_valued_variants_equivalent_to_str():
```
Comment (alalazo) [inline lib/spack/spack/test/variant.py:740]: (minor) Can you add a brief docstring explaining what is tested here (mainly that we can use `*` in directives as an alias to `str`)?

**Review by alalazo — CHANGES_REQUESTED** [review]
A couple of minor requests, but otherwise LGTM

**Review by alalazo — APPROVED** [review]
Thanks!
